exp/operator/controllers/genericprovider.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// genericProvider a generic type for having ability to write shared function for providers
+type genericProvider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   operatorv1.ProviderSpec   `json:"spec,omitempty"`
+	Status operatorv1.ProviderStatus `json:"status,omitempty"`
+}
```

Comment:
Out of interest, did we consider the approach of using an interface rather than doing conversions? We could have used something like we do for the `Conditions` rather than having to perform type conversions

Reply:
Yes, I looked at this option, it will require implementing each function for each provider. If you have a strong opinion against conversions we can switch to interfaces.

Reply:
I don't have a strong opinion per se, but I do wonder if it is easier and more idiomatic to use interfaces (interfaces are how polymorphism works in Go right?). Also, maybe more in keeping with existing code here? @fabriziopandini Did you have any thoughts on conversion to internal type vs interface for this?

Reply:
IMO what we should avoid is to work with unstructured, which is complex and error prone; given that I'm open to any solution.
In this case, I'm not really sure we can use interfaces given that API types mostly expose fields vs exposing methods.

Reply:
My suggested interface would be 
```
type Provider interface {
  client.Object
  GetSpec() operatorv1.ProviderSpec
  SetSpec(in operatorv1.ProviderSpec)
  GetStatus() operatorv1.ProviderStatus
  SetStatus(in operatorv1.ProviderStatus)
}
```

Implementing these 4 simple functions on each of the 4 types of providers would mean that they can all be passed into the functions that need to be generic and instead of reading the spec from the object we would `provider.GetSpec().field`, so it's a simple change without having to do any conversion

This is very much inspired by the Conditions interface that we have across CAPI today